docs/advanced/debugging.rst

```diff
@@ -27,15 +27,335 @@ And, of course, if you're hitting that ever-challenging ``No scope with a Tag ma
+===========
+
+The easiest way to get started with diagnostics is to use the ``Autofac.Diagnostics.DefaultDiagnosticTracer`` class. This tracer will generate a hierarchical trace of resolution operations that you can use in troubleshooting.
+
+.. sourcecode:: csharp
+
+    // Build your container as you normally would.
+    var containerBuilder = new ContainerBuilder();
+    containerBuilder.RegisterType<Component>().As<IService>();
+    var container = containerBuilder.Build();
+
+    // Create a tracer instance and figure out
+    // what you want to do when a trace is ready
+    // to be viewed.
+    var tracer = new DefaultDiagnosticTracer();
+    tracer.OperationCompleted += (sender, args) =>
+    {
+        Trace.WriteLine(args.TraceContent);
+    };
+
+    // Subscribe to the diagnostics with your tracer.
+    container.SubscribeToDiagnostics(tracer);
+
+    // Resolve things as normal. The tracer
+    // event will be raised whenever something
+    // is resolved.
+    using var scope = container.BeginLifetimeScope();
+    scope.Resolve<IService>();
+
+Default Diagnostic Tracer
+=========================
+
+The quick start above shows how you can get going with the ``Autofac.Diagnostics.DefaultDiagnosticTracer``.
+
+When the ``OperationCompleted`` event gets raised, you'll get event arguments that provide:
+
+* ``Operation`` - the actual resolve operation that completed so you can inspect it if needed.
+* ``TraceContent`` - the built string that has the full trace of the resolve operation.
+
+Let's say you have a simple lambda that registers a string.
+
+.. sourcecode:: csharp
+
+    var builder = new ContainerBuilder();
+    builder.Register(ctx => "HelloWorld");
+    var container = builder.Build();
+
+If you resolve a string from that container, the trace will look like this::
+
+    Resolve Operation Starting
+    {
+      Resolve Request Starting
+      {
+        Service: System.String
+        Component: λ:System.String
+
+        Pipeline:
+        -> CircularDependencyDetectorMiddleware
+          -> ScopeSelectionMiddleware
+            -> SharingMiddleware
+              -> RegistrationPipelineInvokeMiddleware
+                -> ActivatorErrorHandlingMiddleware
+                  -> DisposalTrackingMiddleware
+                    -> λ:System.String
+                    <- λ:System.String
+                  <- DisposalTrackingMiddleware
+                <- ActivatorErrorHandlingMiddleware
+              <- RegistrationPipelineInvokeMiddleware
+            <- SharingMiddleware
+          <- ScopeSelectionMiddleware
+        <- CircularDependencyDetectorMiddleware
+      }
+      Resolve Request Succeeded; result instance was HelloWorld
+    }
+    Operation Succeeded; result instance was HelloWorld
+
+As you can see, the trace is very detailed - you can see the full middleware pipeline that the resolve operation went through, you'll see the activator (a delegate, in this case), and you'll see the resulting instance.
```

Comment:
Might it be useful to link to the pipeline docs here, the first time we mention a middleware pipeline? For a lot of users, this diagnostics graph will be the first time they see middleware at all.

Reply:
This moved with the reorder, but the pipeline doc link was added.